app/models/transformation_mapping/vm_migration_validator.rb

```diff
@@ -0,0 +1,141 @@
+class TransformationMapping::VmMigrationValidator
+  require 'miq-hash_struct'
+
+  VM_CONFLICT = "conflict".freeze
+  VM_EMPTY_NAME = "empty_name".freeze
+  VM_IN_OTHER_PLAN = "in_other_plan".freeze
+  VM_INVALID = "invalid".freeze
+  VM_MIGRATED = "migrated".freeze
+  VM_NOT_EXIST = "not_exist".freeze
+  VM_VALID = "ok".freeze
+
+  def initialize(mapping, vm_list = nil)
+    @mapping = mapping
+    @vm_list = vm_list
+  end
+
+  def validate
+    @vm_list.present? ? identify_vms : select_vms
+  end
+
+  def select_vms
+    valid_list = []
+
+    @mapping.transformation_mapping_items.includes(:source => :vms).where(:source_type => EmsCluster).collect(&:source).flat_map(&:vms).each do |vm|
```

Comment:
Sorry, responding late to the ping from @bdunne 's [comment](https://github.com/ManageIQ/manageiq/pull/17364#discussion_r185238164) from before.

I think the suggestion @bdunne did have was pretty decent and simple, and should work for the most part (repeating just for reference):

```ruby
cluster_ids = @mapping.transformation_mapping_items.where(:source_type => EmsCluster).pluck(:source_id)
Vm.where(:ems_cluster_id => cluster_ids).all.each do |vm|
.....
```

Do we possibly want to change `.where(:source_type => EmsCluster)` to something like:

```ruby
.where(:source_type => [EmsCluster] + EmsCluster.descendants)
```

Since there is one subclass of `EmsCluster` (not sure if that matters...).

* * *

My own suggestion would be flipping this query on it's head, and starting from `Vm`, and then you should be able to make only one query:

```ruby
sub_select = @mapping.transformation_mapping_items.select(:source_id)
                                                  .where(:source_type => EmsCluster)
Vm.where(:ems_cluster_id => sub_select)
#  Vm Load (1.2ms)  SELECT "vms".* FROM "vms" WHERE "vms"."type" IN ('Vm', 'ManageIQ::Providers::InfraManager::Vm', 'ManageIQ::Providers::CloudManager::Vm', 'VmServer', 'VmXen', 'ManageIQ::Providers::Kubevirt::InfraManager::Vm', 'ManageIQ::Providers::Redhat::InfraManager::Vm', 'ManageIQ::Providers::Microsoft::InfraManager::Vm', 'ManageIQ::Providers::Vmware::InfraManager::Vm', 'ManageIQ::Providers::Amazon::CloudManager::Vm', 'ManageIQ::Providers::Azure::CloudManager::Vm', 'ManageIQ::Providers::Google::CloudManager::Vm', 'ManageIQ::Providers::Openstack::CloudManager::Vm', 'ManageIQ::Providers::Vmware::CloudManager::Vm') AND "vms"."template" = $1 AND "vms"."ems_id" IN (SELECT "transformation_mapping_items"."source_id" FROM "transformation_mapping_items" WHERE "transformation_mapping_items"."source_type" = 'EmsCluster' AND "transformation_mapping_items"."transformation_mapping_id" = 1) [["template", "f"]]
```

Reply:
This looks great - if you note, @NickLaMuro used a `select` not `pluck` here